atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/DbKvs.java

```diff
@@ -624,46 +629,57 @@ private ColumnSelection getColumnSelectionForRange(RangeRequest range) {
-            cellsByRow = cellsByRow.descendingMap();
-        }
+                Cells.breakCellsUpByRow(Multimaps.asMap(results.results));
+
         List<RowResult<Set<Long>>> finalResults = Lists.newArrayListWithCapacity(results.size());
         for (Entry<byte[], SortedMap<byte[], Set<Long>>> entry : cellsByRow.entrySet()) {
             finalResults.add(RowResult.create(entry.getKey(), entry.getValue()));
         }
-        byte[] nextRow = null;
-        boolean mayHaveMoreResults = false;
-        byte[] lastRow = range.isReverse() ? rows.first() : rows.last();
-        if (!RangeRequests.isTerminalRow(range.isReverse(), lastRow)) {
-            nextRow = RangeRequests.getNextStartRow(range.isReverse(), lastRow);
-            mayHaveMoreResults = rows.size() == maxRows;
+        byte[] nextRow = results.currentRow;
+        boolean mayHaveMoreResults = !results.finishedRange;
+        if (results.isEmpty() && nextRow != null) {
+            if (!RangeRequests.isTerminalRow(range.isReverse(), nextRow)) {
+                mayHaveMoreResults = false;
+            } else {
+                nextRow = RangeRequests.getNextStartRow(range.isReverse(), nextRow);
+            }
         }
         return SimpleTokenBackedResultsPage.create(nextRow, finalResults, mayHaveMoreResults);
     }
 
-    private SetMultimap<Cell, Long> getTimestampsByCell(DbReadTable table,
-                                                        Iterable<byte[]> rows,
-                                                        ColumnSelection columns,
-                                                        long timestamp) {
-        SetMultimap<Cell, Long> results = HashMultimap.create();
-        try (ClosableIterator<AgnosticLightResultRow> rowResults = table.getAllRows(rows, columns, timestamp, false)) {
-            while (rowResults.hasNext()) {
-                AgnosticLightResultRow row = rowResults.next();
-                Cell cell = Cell.create(row.getBytes("row_name"), row.getBytes("col_name"));
-                long ts = row.getLong("ts");
-                results.put(cell, ts);
+    private ResultWithToken getTimestampsByCell(
+            DbReadTable table,
+            Iterable<byte[]> rows,
+            ColumnSelection columns,
+            long timestamp,
+            long batchSize,
+            boolean isReverse) {
+        try (ClosableIterator<AgnosticLightResultRow> iterator = table
+                .getAllRows(rows, columns, timestamp, false, isReverse)) {
+            ResultWithToken result = new ResultWithToken();
+            SetMultimap<Cell, Long> rowBuffer = HashMultimap.create();
+            while (iterator.hasNext() && result.size() + rowBuffer.size() < batchSize) {
+                AgnosticLightResultRow cellResult = iterator.next();
+                result.storePreviousRowIfComplete(cellResult, rowBuffer);
+                Cell cell = Cell.create(cellResult.getBlob("row_name"), cellResult.getBlob("col_name"));
```

Comment:
That's a good point, I'll look into getting the numbers before this gets a green light. I only changed it in the first place because getBytes specifically says it's deprecated and getBlob should be used instead.